### ISSUE TITLE
Optimistic UI Improvements

### ISSUE THREAD

**Issue by theodorDiaconu** [cult-of-coders/redis-oplog#124]
Optimistic UI should work much like a synthetic processor.

Optimistic UI should only work with the current observer, not push the changes to all of them.

Refactoring strategy:
We should allow triggering optimistic-ui changes only to the current observer, or somehow use the optimistic UI from Meteor

**Comment by ramezrafla** [comment]
@theodorDiaconu I noticed that changes seem to flicker until the server makes the final call. Is that what you see too?

**Comment by theodorDiaconu** [comment]
@ramezrafla it shouldn't be the case with the latest version. Maybe you have a diff use-case?

**Comment by ramezrafla** [comment]
@theodorDiaconu I just checked the frames of the WS in the dev console, with or without redis-oplog the frames look exactly the same. It only happens when deployed on server (i.e. not on dev with meteor run)

Yet .. the flickering only happens with redis-oplog in production (likely network latency makes it apparent). It's likely an issue on our side, in Blaze, or yet another edge case. I can't debug on server as there is no source-map.

**Comment by theodorDiaconu** [comment]
@ramezrafla can you paste the frames in prod ?

**Comment by theodorDiaconu** [comment]
Inside a mutation call.
Get the session of the current user.
Send events to the session.
```
             const invoke = DDP._CurrentInvocation.get();
                // const session = Meteor.server.sessions[invoke.connection.id];
                // sendAdded sendChanged sendRemoved
                // console.log(session.added, session.changed, session.removed);
```

**Comment by dnish** [comment]
@theodorDiaconu Just wondering if this issue explains the behavior within our app. We're having 1 method which is defined separately on client and server. For example, this is client side:

client/methods.js

    changeChat: function () {

        Meteor.users.update(Meteor.userId(),{$unset:{'profile.chatId':''}});
    }


server/methods.js

    changeChat: function () {

        [...server logic with database queries...]
    }

We've used this construction to simulate a "Connecting..." message if the user has left the chat and waits until the server found a new one for him.

After switching to Redis Oplog we've detected some flickering within our app. After debugging it I found out, that <code>'profile.chatId':</code> caused the issue. So if we call <code>changeChat</code> on server side, we get the following output for the reactive variable <code>profile.chatId</code>

```
undefined
idOfChatBefore
undefined
newChatId
```

The normal behavior would be

```
undefined
newChatId
```

So it seems like that the client doesn't wait for the response of the server, it unsets the field but immediately reverses the change back to the old value.
